File: mySolutions/200. Number of Islands/200.Number of Islands.java
class Solution {
    public int numIslands(char[][] grid) {

        if (grid == null || grid.length == 0 || grid[0].length == 0) {
            return 0;
        }

        boolean[][] visited = new boolean[grid.length][grid[0].length];
        int count = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (!visited[i][j]) {
                    visited[i][j] = true;
                    if (grid[i][j] == '1') {
                        count++;
                        dfs(grid, visited, i, j);
                    }
                }
            }
        }

        return count;
    }

    private void dfs(char[][] grid, boolean[][] visited, int x, int y) {
        int[] dx = new int[] { 1, 0, -1, 0 };
        int[] dy = new int[] { 0, 1, 0, -1 };

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (!(0 <= nx && nx < grid.length && 0 <= ny && ny < grid[0].length)) {
                continue;
            }

            if (!visited[nx][ny] && grid[nx][ny] == '1') {
                visited[nx][ny] = true;
                dfs(grid, visited, nx, ny);
            }
        }

    }
}

File: mySolutions/0 Amazon_OA/Number of Clusters/200. Number of Islands.py
class Solution:

    LAND = '1'
    WATER = '0'
    VISITED = '-1'
    DIRS = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    def numIslands(self, grid: List[List[str]]) -> int:

        if len(grid) == 0:
            return 0

        count = 0

        for y in range(len(grid)):
            for x in range(len(grid[0])):
                if grid[y][x] == self.LAND:
                    count += 1
                    grid[y][x] = self.VISITED
                    self.dfs(grid, x, y)

        return count

    def dfs(self, grid, x, y):
        for i in range(len(self.DIRS)):
            nx = x + self.DIRS[i][0]
            ny = y + self.DIRS[i][1]

            if self.isLand(grid, nx, ny):
                grid[ny][nx] = self.VISITED
                self.dfs(grid, nx, ny)


    def isLand(self, grid, x, y):
        return x >= 0 and x < len(grid[0]) and y >= 0 and y < len(grid) and grid[y][x] == self.LAND